File: 0350-Intersection-of-Two-Arrays-II/0350.py
from typing import Counter, List


class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        hash = Counter(nums1)
        res = []

        for num in nums2:
            if num in hash.keys() and hash[num] > 0:
                res.append(num)
                hash[num] -= 1
        
        return res

    def intersect1(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1 = sorted(nums1)
        nums2 = sorted(nums2)
        i = 0
        j = 0
        res = []

        while i < len(nums1) and j < len(nums2):
            if nums1[i] < nums2[j]:
                i += 1
            elif nums1[i] > nums2[j]:
                j += 1
            else:  # nums1[i] == nums2[j]
                res.append(nums1[i])
                i += 1
                j += 1
        
        return res

File: 0350-Intersection-of-Two-Arrays-II/0350.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        Map<Integer, Integer> hash = new HashMap<>();
        List<Integer> res = new ArrayList<>();

        for (int num : nums1) {
            hash.put(num, hash.getOrDefault(num, 0) + 1);
        }

        for (int num : nums2) {
            if (hash.containsKey(num) && hash.get(num) > 0) {
                res.add(num);
                hash.put(num, hash.get(num) - 1);
            }
        }

        return res.stream().mapToInt(Integer::intValue).toArray();
    }

    public int[] intersect1(int[] nums1, int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        List<Integer> res = new ArrayList<>();
        int i = 0, j = 0;

        while (i < nums1.length && j < nums2.length) {
            if (nums1[i] < nums2[j]) {
                i++;
            } else if (nums1[i] > nums2[j]) {
                j++;
            } else {
                res.add(nums1[i]);
                i++;
                j++;
            }
        }

        return res.stream().mapToInt(Integer::intValue).toArray();
    }
}